src/Nethermind/Nethermind.HealthChecks/FreeDiskSpaceChecker.cs

```diff
@@ -0,0 +1,111 @@
+//  Copyright (c) 2022 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+//
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Nethermind.Api;
+using Nethermind.Core.MessageBus;
+using Nethermind.Logging;
+
+namespace Nethermind.HealthChecks
+{
+    public readonly struct LowDiskSpaceMessage : IMessage
+    {
+        public LowDiskSpaceMessage(long freeDiskSpace)
+        {
+            AvailableDiskSpce = freeDiskSpace;
+        }
+        public long AvailableDiskSpce { get; init; }
+    }
+
+    internal class FreeDiskSpaceChecker : IHostedService, IAsyncDisposable
+    {
+        private readonly ISimpleMessageBus _messageBus;
+        private readonly IHealthChecksConfig _healthChecksConfig;
+        private readonly IInitConfig _initConfig;
+        private readonly ILogger _logger;
+        private readonly PeriodicTimer _timer;
+        private Task _timerTask;
+        public static readonly int BytesToGB = 1024 << 20;
+
+        public FreeDiskSpaceChecker(INethermindApi api)
+        {
+            _messageBus = api.MessageBus;
+            _healthChecksConfig = api.Config<IHealthChecksConfig>();
+            _initConfig = api.Config<IInitConfig>();
+            _logger = api.LogManager.GetClassLogger();
+
+            _timer = new PeriodicTimer(TimeSpan.FromMinutes(5));
+        }
+
+        private async Task CheckDiskSpace(CancellationToken cancellationToken)
+        {
+            try
+            {
+                while (await _timer.WaitForNextTickAsync(cancellationToken))
+                {
+                    DriveInfo di = new(_initConfig.BaseDbPath);
+                    double freeSpacePcnt = (double)di.AvailableFreeSpace / di.TotalSize * 100;
+                    if (freeSpacePcnt < _healthChecksConfig.LowStorageSpaceShutdownThreshold)
+                    {
+                        if (_logger.IsError)
+                            _logger.Error($"Free disk space is below {_healthChecksConfig.LowStorageSpaceShutdownThreshold:0.00}% - shutting down...");
+                        await _messageBus.Publish(new LowDiskSpaceMessage(di.AvailableFreeSpace));
```

Comment:
A bit unnecessary IMO. I'd inject `ProcessCloseCancellationSource` and just cancel it. Or `Environment.Exit`, which is what's generally used.

Reply:
I don't like `Environment.Exit` as it won't provide a graceful shutdown, which is what I'm trying to achieve. In terms of `ProcessCloseCancellationSource` - yes, I could do that, but would have to extend generic `INethermindPlugin` interface (doesn't feel elegant). Alternatively, I could wrap `HealthChecksPlugin` with a dedicated interface and expose via `NethermindApi` - I've seen some other plugins already exposed like this in `IBasicAPI`. Then I could just use an event and subscribe to it directly and get rid of publish/subscribe classes.

Reply:
All above ideas are seems simpler than the one implemented. We could have ProcessCloseCancellationSource in INethermindApi.

Reply:
Apologies - you are right, it will work, just need to call it from synchronous code not to cause a deadlock - I'll simplify the solution.